pyanaconda/payload/dnfpayload.py

```diff
@@ -412,80 +414,149 @@ def addRepo(self, ksrepo):
-            except payload.NoSuchGroup as e:
-                self._miss(e)
+            log.info("selected env: %s", env)
 
+        # groups from kickstart data
         for group in self.data.packages.groupList:
-            if group.name == 'core' or group.name in excludedGroups:
-                continue
-
             default = group.include in (GROUP_ALL,
                                         GROUP_DEFAULT)
             optional = group.include == GROUP_ALL
 
-            try:
-                self._select_group(group.name, default=default, optional=optional)
-                log.info("selected group: %s", group.name)
-            except payload.NoSuchGroup as e:
-                self._miss(e)
+            # Packages in groups can have different types
+            # and we provide an option to users to set
+            # which types are going to be installed
+            # via the --nodefaults and --optional options.
+            #
+            # To not clash with module definitions we
+            # only use type specififcations if --nodefault,
+            # --optional or both are used
+            if not default or optional:
+                types = "mandatory"
+                if default:
+                    types = "{},default".format(types)
+                if optional:
+                    types = "{},optional".format(types)
+
+                group_spec = "@{group_name}/{types}".format(group_name=group.name, types=types)
+            else:
+                # if group is a regular group this is equal to @group/mandatory,default
+                group_spec = "@{group_name}".format(group_name=group.name)
 
+            group_specs.add(group_spec)
+
+        # handle packages
+        selected_packages = set()
+        excluded_packages = set()
         for pkg_name in self.data.packages.excludedList:
-            self._exclude_package(pkg_name)
             log.info("excluded package: '%s'", pkg_name)
+            excluded_packages.add(pkg_name)
 
         for pkg_name in self.data.packages.packageList:
-            try:
-                self._install_package(pkg_name)
-                log.info("selected package: '%s'", pkg_name)
-            except payload.NoSuchPackage as e:
-                self._miss(e)
+            log.info("selected package: '%s'", pkg_name)
+            selected_packages.add(pkg_name)
+
+        kernel_package = self._get_kernel_package()
+        if kernel_package:
+            selected_packages.add(kernel_package)
+
+        # resolve packages and groups required by Anaconda
+        self.requirements.apply()
 
-        self._select_kernel_package()
+        # let's piece it all together and feed it to DNF at once
+        include_set = set()
+        exclude_set = set()
+        # note about formatting:
+        # - leading @ signifies a group or module
+        # - no leading @ means a package
+
+        # add environment
+        if env:
+            include_set.add("@^{}".format(env))
```

Comment:
Can we create this specification when we choose the environment?

Reply:
Done.